### ISSUE TITLE
Filtered posts but no filters set up

### ISSUE THREAD

**Issue by WelshPixie** [mastodon/mastodon#25952]
### Steps to reproduce the problem

Browsing the Local feed, saw a post behind a filter but I have no filters set up. 

### Expected behaviour

The post would be shown without the filter

### Actual behaviour

The post was shown behind a filter

### Detailed description

I don't have any filters set up, but a post appeared behind a filter in my feed that I had to click through to see the post. This was on mastodon.art on the Curator account, we're on 4.1.4.

### Specifications

Mastodon 4.1.4

**Comment by selfawaresoup** [comment]
The same has recently happened on our instance (4.1.0+glitch, 4.1.1+glitch, and 4.1.2+glitch)

A user who has not set up any keyword filters has reported seeing several filtered posts on the local timeline.

**Comment by brookmiles** [comment]
On the flip side, I *do* have filters set up, but I noticed at least one occurrence where I had the Local view open, and my filter failed to apply to a streamed post. When I refreshed the page, the filter applied correctly.

Potentially a separate problem, but if the underlying issue is "the wrong filters get applied to streaming posts", it would make sense that it could fail in this way as well.

**Comment by EeveeEuphoria** [comment]
I can also confirm this on behalf of many of my friends, I've seen many screenshots and experiences of this happening on both vanilla Mastodon and it's various forks.

**Comment by ClearlyClaire** [comment]
Thanks for the report, I'll try to track it down.

I currently don't know what is happening, but a similar issue has been fixed a while ago (#20719, in 4.0.2). I assume people affected are not running that old version of the streaming server, but it would be worth double-checking.

**Comment by selfawaresoup** [comment]
We were running the streaming service at the same version as the main web process when the bug happened, as we always do

**Comment by EeveeEuphoria** [comment]
this also happens on yiff.life, in which the instance my friend saw of this bug was indeed on v4.1.4

**Comment by ClearlyClaire** [comment]
I am still unable to reproduce or identify the issue, but if you're affected, try applying this patch and watching for “Cached filters were already set”, “Account filter mismatch” and “Using mismatched filter” warnings in the logs:
```patch
diff --git a/streaming/index.js b/streaming/index.js
index ffb570c5b9..0814b7ebe5 100644
--- a/streaming/index.js
+++ b/streaming/index.js
@@ -359,6 +359,9 @@ const startWorker = async (workerId) => {
         req.accountId = result.rows[0].account_id;
         req.chosenLanguages = result.rows[0].chosen_languages;
         req.deviceId = result.rows[0].device_id;
+        if (req.cachedFilters) {
+          log.warn('Cached filters were already set');
+        }
 
         resolve();
       });
@@ -690,7 +693,7 @@ const startWorker = async (workerId) => {
         }
 
         if (!unpackedPayload.filtered && !req.cachedFilters) {
-          queries.push(client.query('SELECT filter.id AS id, filter.phrase AS title, filter.context AS context, filter.expires_at AS expires_at, filter.action AS filter_action, keyword.keyword AS keyword, keyword.whole_word AS whole_word FROM custom_filter_keywords keyword JOIN custom_filters filter ON keyword.custom_filter_id = filter.id WHERE filter.account_id = $1 AND (filter.expires_at IS NULL OR filter.expires_at > NOW())', [req.accountId]));
+          queries.push(client.query('SELECT filter.id AS id, filter.phrase AS title, filter.context AS context, filter.expires_at AS expires_at, filter.action AS filter_action, keyword.keyword AS keyword, keyword.whole_word AS whole_word, filter.account_id AS account_id FROM custom_filter_keywords keyword JOIN custom_filters filter ON keyword.custom_filter_id = filter.id WHERE filter.account_id = $1 AND (filter.expires_at IS NULL OR filter.expires_at > NOW())', [req.accountId]));
         }
 
         Promise.all(queries).then(values => {
@@ -704,10 +707,15 @@ const startWorker = async (workerId) => {
             const filterRows = values[accountDomain ? 2 : 1].rows;
 
             req.cachedFilters = filterRows.reduce((cache, row) => {
+              if (req.accountId !== row.account_id) {
+                log.warn(`Account filter mismatch: ${req.accountId} VS ${row.account_id}`);
+              }
+
               if (cache[row.id]) {
                 cache[row.id].keywords.push([row.keyword, row.whole_word]);
               } else {
                 cache[row.id] = {
+                  accountId: row.account_id,
                   keywords: [[row.keyword, row.whole_word]],
                   expires_at: row.expires_at,
                   repr: {
@@ -751,6 +759,10 @@ const startWorker = async (workerId) => {
             const now = new Date();
             payload.filtered = [];
             Object.values(req.cachedFilters).forEach((cachedFilter) => {
+              if (cachedFilter.accountId && cachedFilter.accountId !== req.accountId) {
+                log.warn(`Using mismatched filter: ${req.accountId} VS ${cachedFilter.accountId}`);
+              }
+
               if ((cachedFilter.expires_at === null || cachedFilter.expires_at > now)) {
                 const keyword_matches = searchIndex.match(cachedFilter.regexp);
                 if (keyword_matches) {
```

**Comment by Taullo** [comment]
Alright I'm back after doing some tests 😤​ 

I applied the patch but there's no difference in the log (watched in real-time and even searched "mismatch" a few times) but the bug still occurs.

I set up a filter for "the" on one account and a filter for "and" on another account and watched the firehose page from both. It was behaving extremely strangely. With both accounts monitoring the firehose page at the same time, sometimes one account's filter would show up, sometimes they both would, and sometimes it wouldn't filter at all.

As far as I can tell, I think it only applies the wrong filter when both accounts are looking at the firehose page. It also only applies to new posts loading in, not ones already loaded at page load. I thought it was a CDN caching bug at first but I checked with another instance that isn't behind a CDN and the same behavior happens.

Also, just to rule it out, I threw a dummy log message into streaming/index just to make sure it was loading the correct streaming file with the diff applied (I don't know why it wouldn't be but you never know asdlfjsdf) and it logged the dummy message fine.
